hedera-node/hedera-app/src/main/java/com/hedera/node/app/info/DiskStartupNetworks.java

```diff
@@ -162,67 +167,133 @@ public static void writeNetworkInfo(@NonNull final State state, @NonNull final P
+        requireNonNull(platformStateStore);
+        Optional.ofNullable(rosterStore.getActiveRoster())
+                .or(() -> Optional.ofNullable(buildRoster(platformStateStore.getAddressBook())))
+                .ifPresent(activeRoster -> {
+                    final var network = Network.newBuilder();
+                    final List<NodeMetadata> nodeMetadata = new ArrayList<>();
+                    activeRoster.rosterEntries().forEach(entry -> {
+                        final var node = requireNonNull(nodeStore.get(entry.nodeId()));
+                        final var encryptionKey = Optional.ofNullable(tssStore.getTssEncryptionKeys(node.nodeId()))
+                                .map(TssEncryptionKeys::currentEncryptionKey)
+                                .orElse(Bytes.EMPTY);
+                        nodeMetadata.add(new NodeMetadata(entry, node, encryptionKey));
+                    });
+                    network.nodeMetadata(nodeMetadata);
+                    final var currentRosterHash = rosterStore.getCurrentRosterHash();
+                    if (currentRosterHash != null) {
+                        final var sourceRosterHash = Optional.ofNullable(rosterStore.getPreviousRosterHash())
+                                .orElse(Bytes.EMPTY);
+                        tssStore.consensusRosterKeys(sourceRosterHash, currentRosterHash, rosterStore)
+                                .ifPresent(rosterKeys ->
+                                        network.ledgerId(rosterKeys.ledgerId()).tssMessages(rosterKeys.tssMessages()));
+                    }
+                    try (final var fout = Files.newOutputStream(path)) {
+                        Network.JSON.write(network.build(), new WritableStreamingData(fout));
+                    } catch (IOException e) {
+                        log.warn("Failed to write network info", e);
+                    }
+                });
+    }
+
+    /**
+     * Converts a {@link AddressBook} to a {@link Network}. The resulting network will have no TSS
+     * keys of any kind.
+     *
+     * @param addressBook the address book to convert
+     * @return the converted network
+     */
+    public static @NonNull Network fromLegacyAddressBook(@NonNull final AddressBook addressBook) {
+        final var roster = buildRoster(addressBook);
+        return Network.newBuilder()
+                .nodeMetadata(roster.rosterEntries().stream()
+                        .map(rosterEntry -> {
+                            final var nodeId = rosterEntry.nodeId();
+                            final var nodeAccountId = parseAccount(
+                                    addressBook.getAddress(NodeId.of(nodeId)).getMemo());
+                            // Currently the ReadableFreezeUpgradeActions.writeConfigLineAndPem()
+                            // assumes that the gossip endpoints in the Node objects are in the order
+                            // (Internal, External)...even though Roster format is the reverse :/
```

Comment:
Agreed. I am not sure why it is reverse in Roster. It seems like that should be fixed to avoid all the places we need to make sure that is correct.

Reply:
The Roster is a data structure used by Consensus, and we require the external endpoint to be at index zero. In fact, we plan to only ever have external endpoints listed in the Roster. Internal endpoints don't belong to the Roster because it's a public data structure. Internal endpoints belong to a node-specific local config. @edward-swirldslabs  is currently working on implementing this. In the mean time, the Roster must continue to have the external endpoint at index zero.